src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,412 @@
+      podDefinition,
+      instanceId,
+      container,
+      offer.getHostname,
+      resourceMatch.hostPorts,
+      envPrefix)
+
+    builder.setCommand(commandInfo)
+
+    val containerInfo = computeContainerInfo(podDefinition.podVolumes, container)
+
+    builder.setContainer(containerInfo)
+
+    container.healthCheck.foreach { healthCheck =>
+      builder.setHealthCheck(computeHealthCheck(healthCheck, container.endpoints))
+    }
+
+    builder
+  }
+
+  private[this] def computeExecutorInfo(
+    podDefinition: PodDefinition,
+    instanceId: Instance.Id): mesos.ExecutorInfo.Builder = {
+    val executorID = mesos.ExecutorID.newBuilder.setValue(f"marathon-${instanceId.idString}")
+
+    val executorInfo = mesos.ExecutorInfo.newBuilder
+      .setType(mesos.ExecutorInfo.Type.DEFAULT)
+      .setExecutorId(executorID)
+
+    // We're deliberately not setting executor resources here. Because we're using the default executor, the
+    // default resource values can be used.
+
+    if (podDefinition.networks.nonEmpty) {
+      val containerInfo = mesos.ContainerInfo.newBuilder
+        .setType(mesos.ContainerInfo.Type.MESOS)
+
+      // TODO: What about host networks? Is this only for labels?
+      // TODO: Does 'DiscoveryInfo' need to be set?
+
+      podDefinition.networks
+        .filter(network => network.name.isDefined && network.mode == Some(raml.NetworkMode.Container))
+        .foreach{ hostNetwork =>
+          val networkInfo = mesos.NetworkInfo.newBuilder
+            .setName(hostNetwork.name.get)
+
+          hostNetwork.labels.foreach { labels =>
+            networkInfo.setLabels(
+              mesos.Labels.newBuilder.addAllLabels(labels.values.map {
+                case (key, value) =>
+                  mesos.Label.newBuilder.setKey(key).setValue(value).build
+              }.asJava))
+          }
+
+          containerInfo.addNetworkInfos(networkInfo)
+        }
+
+      executorInfo.setContainer(containerInfo)
+    }
+
+    if (podDefinition.labels.nonEmpty) {
+      val labels = mesos.Labels.newBuilder
+        .addAllLabels(podDefinition.labels.map {
+          case (key, value) =>
+            mesos.Label.newBuilder.setKey(key).setValue(value).build
+        }.asJava)
+
+      executorInfo.setLabels(labels)
+    }
+
+    executorInfo
+  }
+
+  private[this] def computeCommandInfo(
+    podDefinition: PodDefinition,
+    instanceId: Instance.Id,
+    container: raml.MesosContainer,
+    host: String,
+    hostPorts: Seq[Option[Int]],
+    envPrefix: Option[String]): mesos.CommandInfo.Builder = {
+    val commandInfo = mesos.CommandInfo.newBuilder
+
+    container.exec.foreach{ exec =>
+      exec.command match {
+        case raml.ShellCommand(shell) =>
+          commandInfo.setShell(true)
+          commandInfo.setValue(shell)
+        case raml.ArgvCommand(argv) =>
+          commandInfo.setShell(false)
+          if (argv.nonEmpty) {
+            commandInfo.setValue(argv.head)
```

Comment:
should check exec.overrideEntrypoint. the idea was that if overrideEntypoint == false then argv then is interpreted the parameter list for some entrypoint (so setValue would not be called). if overrideEntrypoint == true then setValue(argv[0]). In any case addAllArguments probably needs the entire argv list, not just the tail.